### ISSUE TITLE
format string enhancement

### ISSUE THREAD

**Issue by Kyle-Kyle** [Gallopsled/pwntools#997]
fmtstr tools place pointers before formats in payloads. 
But such payloads can be useless when vulnerable functions are something like `printf` or `sprintf`. They will stop at `\x00` which often appear in pointers.
If formats are placed before pointers, such problems can be easily resolved since formats are printable.
For example, `'\x41\x41\x41\x41\x41\x7f\x00\x00%100c%10$hhn'` is what payload looks like now.
If we call `printf`, nothing happens.
If we use `'%100c%10$hhn\x41\x41\x41\x41\x41\x7f\x00\x00'`. Good, something is injected.
(well, in the above example, it is supposed that the offset is correct. But you get the idea.)

**Comment by zachriggle** [comment]
See #827

**Comment by Kyle-Kyle** [comment]
Well..Actually, I have reconstructed `libformatstr`. And it is bug free(maybe?) and has more features now. New features  include 64bit support and controlling number of bytes to write. If you wish, you can take it as a reference. Or maybe, can I take this project? I like challenges.

**Comment by zachriggle** [comment]
If you'd like to pick up #827 and continue the work there, that would be much preferred instead of:

1. Updating the current code
2. Replacing the current code with `libformatstr`
3. Replacing the current code with a modified / updated version of `libformatstr`

**Comment by Kyle-Kyle** [comment]
I've gone through the code in formatstring.py. I found that an optimization is applied to memory chunks in order to reduce print length.
But this optimization is one of the buggy parts. Example: 
~~~~
f[0x6cb010] = '\x00\x41'
f[0x6cb011] = '\x41\x00'

outcome:
[write(0x6cb010, 0x4100)]
~~~~
And let's forget about the bug which is relatively easy to fix . Let's talk about the outcome of this optimization. The expected outcome is that the length of our payload is not much affected while print length is hugely reduced. But unfortunately, after some thinking, I found the algorithm not suitable here.
~~~~
f[0x6cb010] = '\x00\x41\x41\x00'
f[0x6cb014] = '\x41\x41'
f[0x6cb018] = '\x41\x41'
~~~~
Forget about the bug, the expected outcome of this optimization is 5 pointers, namely: `0x6cb011`, `0x6cb014`, `0x6cb018`, `0x6cb010`, `0x6cb013`. The simple split-up algorithm gives only 4 pointers, namely: `0x6cb010`, `0x6cb012`, `0x6cb014`, `0x6cb018`. In some cases, this difference can be huge. In case of small buffer size, this situation becomes worse.

In conclusion, this optimization reduce print length which is a minor problem compared with length of payload since we can reduce print length by other means.
So I want to ask for your consent to remove this optimization since it seems that you don't want me to do some big changes.

BTW, I'm considering adding a new feature: force writes in order. Sometimes we can execute vulnerable function only for once. In that case, ordering of writes becomes fatal. For example, overwriting `stdout` should always take place after fake vtable is created.
What do you think about this feature?

**Comment by zachriggle** [comment]
> But this optimization is one of the buggy parts

Yep! It's all work-in-progress code.  However, it's a good foundation for having semi-automated format string exploitation that works with *generic* code on *all* architectures and ABIs, against *any* `*printf` routine.

> The expected outcome is that the length of our payload is not much affected while print length is hugely reduced.

No, you've got the intent backwards.  See the actual [documentation of the optimization](https://github.com/zachriggle/pwntools/blob/4ca6b0bb3f574ed7f20ed385f633df5f9e990f40/pwnlib/formatstring.py#L199-L203):

```
            >>> f[0xdeadbeef] = 'AA'
            >>> f.payload()
            '%16705c%08$hnXXX\xf0\xbe\xad\xde'

...

            Repeated values are optimized and collapsed.

            >>> f[0xcafebabe] = 'AA'
            >>> f.payload()
            '%16705c%08$hn%hn\xf0\xbe\xad\xde\xbf\xba\xfe\xca'
```

> optimization reduce print length which is a minor problem compared with length of payload

The intent is to minimize the length of the format string, not print length.  The code may be busted, currently -- note that it's not merged yet <3

**Comment by zachriggle** [comment]
For the example that you call out:

```py
f[0x6cb010] = '\x00\x41\x41\x00'
f[0x6cb014] = '\x41\x41'
f[0x6cb018] = '\x41\x41'
```

... I'd expect the optimization to only emit four writes ...

```
0x6cb010 --> 00 41
0x6cb012 --> 41 00
0x6cb014 --> 41 41
0x6cb018 --> 41 41
```

As I said at the end of the above comment, this code is all work-in-progress :)

In its current state, it probably makes a poor choice and splits `00 41 41 00` into `00` `41 41` and `00` again, which makes five writes.

I believe this is the same conclusion that you made in your comment -- I'm just trying to show that I agree with you and we're on the same page.

The code just isn't done yet <3

**Comment by iromise** [comment]
Happy to see that this feature is on the way. Is there anything I can do? Hope it can be used soonly.

@Kyle-Kyle  I am wondering whether you are still developing this feature?

**Comment by zachriggle** [comment]
Feel free to fork `pwntools` and work on my branch.  Hopefully we can get things working.

I don't have any cycles to dedicate to this currently.

Closing as a duplicate of #827.

The most recent status is in my comment here: https://github.com/Gallopsled/pwntools/pull/827#issuecomment-350085470
